File: 프로그래머스/lv4/118670. 행렬과 연산/행렬과 연산.java
import java.util.ArrayDeque;

class Solution {
    public int[][] solution(int[][] rc, String[] operations) {
        int[][] answer = new int[rc.length][rc[0].length];
        
        ArrayDeque<Integer> left = new ArrayDeque<>();
        ArrayDeque<Integer> right = new ArrayDeque<>();
        ArrayDeque<ArrayDeque<Integer>> mid = new ArrayDeque<>();
        
        for(int[] a : rc){
            ArrayDeque<Integer> row = new ArrayDeque<>();
            for(int j = 0; j < a.length; j++){
                if(j == 0){
                    left.add(a[j]);
                }
                else if(j == a.length - 1){
                    right.add(a[j]);
                }
                else{
                    row.add(a[j]);
                }
            }
            mid.add(row);
        }
        
        for(String operation : operations){
            if("ShiftRow".equals(operation)){
                left.addFirst(left.pollLast());
                mid.addFirst(mid.pollLast());
                right.addFirst(right.pollLast());
            }
            else{
                mid.peekFirst().addFirst(left.pollFirst());
                right.addFirst(mid.peekFirst().pollLast());
                mid.peekLast().addLast(right.pollLast());
                left.addLast(mid.peekLast().pollFirst());
            }
        }
        
        for(int i = 0; i < rc.length; i++){
            int j = 0;
            answer[i][j++] = left.pollFirst();
            ArrayDeque<Integer> row = mid.pollFirst();
            while(!row.isEmpty()){
                answer[i][j++] = row.pollFirst();
            }
            answer[i][j] = right.pollFirst();
        }
        
        return answer;
    }
}

File: 프로그래머스/4/118670. 행렬과 연산/행렬과 연산.cpp
#include <string>
#include <vector>
#include <deque>

using namespace std;

vector<vector<int>> solution(vector<vector<int>> rc, vector<string> operations) {

	deque<int> left, right;
	deque<deque<int>> mid;

	int row = rc.size(), col = rc[0].size();
	for (int i = 0; i < row; i++) {
		left.emplace_back(rc[i][0]);
		right.emplace_back(rc[i][col - 1]);
		deque<int> dq;
		for (int j = 1; j < col - 1; j++) {
			dq.emplace_back(rc[i][j]);
		}
		mid.emplace_back(dq);
	}

	for (const auto& oper : operations) {
		if (oper == "ShiftRow") {
			
			left.push_front(std::move(left.back()));
			left.pop_back();

			right.push_front(std::move(right.back()));
			right.pop_back();

			mid.emplace_front(std::move(mid.back()));
			mid.pop_back();
			

		}
		else {
			
			mid[0].emplace_front(std::move(left.front()));
			left.pop_front();

			right.push_front(std::move(mid[0].back()));
			mid[0].pop_back();

			mid[row - 1].emplace_back(std::move(right.back()));
			right.pop_back();

			left.push_back(std::move(mid[row-1].front()));
			mid[row - 1].pop_front();

		}
	}

	vector<vector<int>> answer(row, vector<int>(col, 0));
	for (int i = 0; i < row; i++) {
		answer[i][0] = left[i];
		answer[i][col - 1] = right[i];
		for (int j = 1; j < col - 1; j++) {
			answer[i][j] = mid[i].front();
			mid[i].pop_front();

		}
	}

	return answer;
}